dask_expr/_expr.py

```diff
@@ -60,6 +61,35 @@ def __init__(self, *args, **kwargs):
                 # Raise a ValueError instead of AttributeError to
                 # avoid infinite recursion
                 raise ValueError(f"{dep} has no attribute {self._required_attribute}")
+        self._dependents = []
+
+    def _clear_dependents(self):
+        self._dependents = []
+
+    def _register_dependents(self):
+        for node in self.walk():
+            node._clear_dependents()
+
+        stack = [self]
+        seen = set()
+        while stack:
+            node = stack.pop()
+            if node._name in seen:
+                continue
+            seen.add(node._name)
+
+            for dep in node.dependencies():
+                stack.append(dep)
+                dep._add_dependents(node)
```

Comment:
Maybe a dumb idea, but we could also put this in `__init__` but only use it if we're in an optimization phase or something.  Maybe we turn it on when we enter optimize, and then recreate the full expression tree as a first stage.  Then everything will be kept up to date automatically during optimization and there's no need to explicitly refresh it.  

But again, maybe a dumb idea.

Reply:
Yeah I thought about this too for a little bit, I am not sure if we can ensure that a new node always gets all dependents of the old node right away without moving the logic into simplify (which is something I am trYing to avoid). I was running into too over-eager pushdowns when I tried this